compiler/rustc_typeck/src/astconv/generics.rs

```diff
@@ -43,23 +43,62 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             }
         }
 
+        let add_braces_suggestion = |arg: &GenericArg<'_>, err: &mut DiagnosticBuilder<'_>| {
+            let suggestions = vec![
+                (arg.span().shrink_to_lo(), String::from("{ ")),
+                (arg.span().shrink_to_hi(), String::from(" }")),
+            ];
+            err.multipart_suggestion(
+                "if this generic argument was intended as a const parameter, \
+                   surround it with braces:",
+                suggestions,
+                Applicability::MaybeIncorrect,
+            );
+        };
+
         // Specific suggestion set for diagnostics
         match (arg, &param.kind) {
             (
-                GenericArg::Type(hir::Ty { kind: hir::TyKind::Path { .. }, .. }),
-                GenericParamDefKind::Const { .. },
+                GenericArg::Type(hir::Ty {
+                    kind: hir::TyKind::Path(rustc_hir::QPath::Resolved(_, path)),
+                    ..
+                }),
+                GenericParamDefKind::Const,
             ) => {
-                let suggestions = vec![
-                    (arg.span().shrink_to_lo(), String::from("{ ")),
-                    (arg.span().shrink_to_hi(), String::from(" }")),
-                ];
-                err.multipart_suggestion(
-                    "if this generic argument was intended as a const parameter, \
-                try surrounding it with braces:",
-                    suggestions,
-                    Applicability::MaybeIncorrect,
-                );
+                use rustc_hir::def::{DefKind, Res};
+                match path.res {
+                    Res::Err => {
+                        add_braces_suggestion(arg, &mut err);
+                        err.set_primary_message(
+                            "unresolved item provided when a constant was expected",
+                        );
+                    }
+                    Res::Def(DefKind::TyParam, src_def_id) => (|| {
```

Comment:
Why are you putting this in a closure and calling immediately?

Reply:
immediately invoked function expressions are a pattern I stole from JS, just because I wanted to not execute the rest in the case the branch for checking if its local is hit

Reply:
That's not necessary in Rust.